### PR TITLE
[DDAN-99] High Ability

### PR DESCRIPTION
**Pull request by Sangmin0728**
### PURPOSE OF PR
* Simple summary of this PR
This PR is about the high ability of the master and task_manager.

### What type of PR is it?
* fix / feature / improvement /  hot-fix / refactoring
Improvement

### JIRA ISSUE
* https://ddangcoonfire.atlassian.net/jira/software/c/projects/DDAN/boards/1?modal=detail&selectedIssue=DDAN-98

### HAVE YOU CHECKED ...
* [x] ADD JIRA ID ON PR TITLE (ex. [DDAN-1234] PR TITLE)
* [x] pre-commit
* [ ] Branch Format ( type/issue-NUM ) (ex. refactoring/DDAN-1234 )

### Explanation
( Adding screenshot would be great help )
1.  If one of the task_manager is abnormal, then the master re-allocate the tasks were assigned to the task_manager to other task_managers.
2.  If the master allocate the tasks but there is no available task_manager, then master save the tasks and auto-restart the tasks when one of the task_managers can be used.
3.  By changing the order of threads starting, fix the health check error.
4.  By improving the condition, Process Monitor can communicate to the normal task_manager only.

### CODE REVIEW

File: experiment_scheduler/master/master.py [hunk 1]
```diff
@@ -70,10 +70,13 @@ def __init__(self):
         # [Todo] Logging required
         # [TODO] need discussion about path and env vars
         self.task_managers_address: list = self.get_task_managers()
+        self.retry_task_list = []
```
Comment (Sangmin0728) [inline experiment_scheduler/master/master.py:73]: We use retry_task_list queue for saving retry task when there is no available task_manager.
But we delete this queue and change the logic of execute task.

File: experiment_scheduler/master/master.py [hunk 2]
```diff
@@ -94,6 +97,25 @@ def get_task_managers() -> List[str]:
...
+            if self.process_monitor.selected_task_manager != -1:
+                unhealthy_task_manager_list = []
+                for task_manager in self.task_managers_address:
+                    if not self.process_monitor.thread_queue[f"is_{task_manager}_healthy"]:
+                        unhealthy_task_manager_list.append(task_manager)
+                if len(unhealthy_task_manager_list) > 0:
+                    for unhealthy_task_manager in unhealthy_task_manager_list:
+                        self._retry_execute_task(unhealthy_task_manager)
+            time.sleep(interval)
+
```
Comment (Sangmin0728) [inline experiment_scheduler/master/master.py:118]: Master check the status-info of task_managers.
If there are abnormal task_managers, then master retry execute task.

File: experiment_scheduler/master/master.py [hunk 3]
```diff
@@ -102,19 +124,46 @@ def _execute_command(self, interval=1) -> None:
...
+                    status=TaskStatus.Status.NOTSTART, order_by=TaskEntity.updated_at
                 )
-                if available_task_managers:
-                    task_manager_address = available_task_managers[0]
-                    task_manager = TaskManagerEntity.get(address=task_manager_address)
-                    self.execute_task(task_manager.address, queue_task.id)
+                if queue_task is not None:
+                    self._allocate_task(queue_task)
+            else:
+                self._allocate_task(self.retry_task_list.pop(0), retry=True)
```
Comment (Sangmin0728) [inline experiment_scheduler/master/master.py:134]: Priority of the executing task : Retry > Nonstart.

File: experiment_scheduler/master/master.py [hunk 4]
```diff
@@ -102,19 +124,46 @@ def _execute_command(self, interval=1) -> None:
...
+        available_task_managers = (
+            self.process_monitor.get_available_task_managers()
+        )
+        if available_task_managers:
+            task_manager_address = available_task_managers[0]
+            task_manager = TaskManagerEntity.get(address=task_manager_address)
+            self.execute_task(task_manager.address, task.id)
+        elif retry:
+            if task not in self.retry_task_list:
+                self.retry_task_list.insert(0, task)
```
Comment (Sangmin0728) [inline experiment_scheduler/master/master.py:153]: Refactoring the repeated code by extracting the method.

File: experiment_scheduler/master/process_monitor.py [hunk 5]
```diff
@@ -32,6 +33,7 @@ class ProcessMonitor:
     """
 
     def __init__(self, task_managers: List[str]):
+        self.selected_task_manager = -1
```
Comment (Sangmin0728) [inline experiment_scheduler/master/process_monitor.py:36]: This parameter is the flag that task_manager is not connected whenever.

File: experiment_scheduler/master/process_monitor.py [hunk 6]
```diff
@@ -61,34 +63,40 @@ def _get_stubs(self) -> Dict[str, Any]:
...
-                    )
+                    if self.selected_task_manager != -1:
+                        thread_queue[f"is_{task_manager}_healthy"] = False
+                        self.logger.error(
+                            "currently task manager %s is not available.\n error log : %s",
+                            task_manager,
+                            error,
+                        )
+                    else:
+                        self.logger.warning("task managers are not started yet")
```
Comment (Sangmin0728) [inline experiment_scheduler/master/process_monitor.py:95]: Process monitor check the task_managers with status and finished tasks.
And we change the logger message when task_managers is not started and when task_manager is abnormal.

File: experiment_scheduler/master/process_monitor.py [hunk 7]
```diff
@@ -61,34 +63,40 @@ def _get_stubs(self) -> Dict[str, Any]:
...
-    def _are_task_manager_healthy(self) -> bool:
-        """
-
-        :return:
-        """
-        for address in self.task_manager_address:
-            if not self.thread_queue[f"is_{address}_healthy"]:
-                return False
-        return True
-
```
Comment (Sangmin0728) [inline experiment_scheduler/master/process_monitor.py:91]: It is no used method.

File: experiment_scheduler/master/process_monitor.py [hunk 8]
```diff
@@ -135,6 +143,6 @@ def get_available_task_managers(self):
         """
         available_task_managers = []
         for tm_address, tm_stub in self.task_manager_stubs.items():
-            if tm_stub.has_idle_resource(PROTO_EMPTY).exists:
+            if self.thread_queue[f"is_{tm_address}_healthy"] and tm_stub.has_idle_resource(PROTO_EMPTY).exists:
```
Comment (Sangmin0728) [inline experiment_scheduler/master/process_monitor.py:146]: Process Monitor check the resource of the healthy task_manager only.

File: experiment_scheduler/task_manager/task_manager_server.py [hunk 9]
```diff
@@ -130,8 +130,22 @@ def use_gpu(self):
...
+        done_task_id_list = []
+        if len(task_id_list) > 0:
+            for task_id in task_id_list:
+                task_status = self._get_task_status_by_task_id(task_id)
+                if task_status.status == TaskStatus.Status.DONE:
+                    server_status.task_id_array.append(task_id)
+                    done_task_id_list.append(task_id)
+        for done_task_id in done_task_id_list:
+            del self.tasks[done_task_id]
+        return server_status
```
Comment (Sangmin0728) [inline experiment_scheduler/task_manager/task_manager_server.py:148]: Task Manager returns the status and finished tasks list.

**Review by sinunu — APPROVED** [review]

**Pull request merged**